### PR TITLE
Use traits to determine allocator propagation

### LINKED ISSUE #66: Use allocator traits for allocator propagation

### PR DESCRIPTION
**Pull request by jbcoe**
Fixes #66

### CODE REVIEW

**Comment by jbcoe** [comment]
Complicated merge to do - I'll sort this

**Comment by jbcoe** [comment]
I’ll resolve the merge conflicts once I’m home

File: indirect.h [hunk 1]
```diff
@@ -140,31 +144,57 @@ class indirect {
...
-    } else {
-      indirect tmp(other);
-      this->swap(tmp);
     }
     return *this;
   }
 
-  constexpr indirect& operator=(indirect&& other) noexcept {
+  constexpr indirect& operator=(indirect&& other) noexcept(
+      allocator_traits::propagate_on_container_move_assignment::value) {
```
Comment (Twon) [inline indirect.h:174]: I belive this should also be conditional with or of allocator traits is_always_equal as is the case for std::vector swap https://eel.is/c++draft/vector.capacity
```suggestion
      allocator_traits::propagate_on_container_move_assignment::value || allocator_traits::is_always_equal::value) {
```

Comment (jbcoe) [reply indirect.h:174]: Agreed. Can you make the change?

Comment (Twon) [reply indirect.h]: Yes, just pushed up.  I'm working on the swap methods now

Comment (Twon) [reply indirect.h]: Interestingly it looks like no major STL implementations support the case where propagate_on_container_move_assignment is false or allocator are not equal:

Microsoft: https://github.com/microsoft/STL/blob/f362f7d7bea87bd199c927a89718e929ed89b187/stl/inc/vector#L3452

libstdc++: https://github.com/gcc-mirror/gcc/blob/9a82cca91dee190d845ee1681440bc69a07c6d49/libstdc%2B%2B-v3/include/bits/stl_vector.h#L1591

libc++: https://github.com/llvm/llvm-project/blob/5d35273a32d239b7407338e13ed71b59174d6536/libcxx/include/vector#L2017

**Review by Twon — APPROVED** [review]
The change match my expectation based on my understanding of this via: https://foonathan.net/2015/10/allocatorawarecontainer-propagation-pitfalls/

I've now extended support for POCS also.

**Pull request merged**